packages/core/database/lib/entity-manager/relations-orderer.js

```diff
@@ -1,6 +1,106 @@
+  // Map to store the first index where a relation id is connected
+  const firstSeen = {};
+  // Map to validate if connect relation has already been computed
+  const computedIdx = {};
+
+  connectArr.forEach((rel, idx) => {
+    // If adjacent relation is not in the database or seen yet in the connect array
+    // then we need to sort the connect array
+    const adjacentRelId = rel.position?.before || rel.position?.after;
+    if (!relInArray[adjacentRelId] && !firstSeen[adjacentRelId]) needsSorting = true;
+    // Populate firstSeen
+    if (!(rel.id in firstSeen)) firstSeen[rel.id] = idx;
+  });
+
+  // If we don't need to sort the connect array, we can return it as is
+  if (!needsSorting) return connectArr;
+
+  // Add relation to sortedConnect and mark it as seen
+  const pushRelation = (rel) => {
+    sortedConnect.push(rel);
+    relInArray[rel.id] = true;
+  };
+
+  // Recursively compute in which order the relation should be connected
+  const computeRelation = (rel, idx, relationsSeenInBranch) => {
+    const adjacentRelId = rel.position?.before || rel.position?.after;
+
+    // This connect has already been computed
+    if (idx in computedIdx) return;
+
+    if (!adjacentRelId || relInArray[adjacentRelId]) {
+      return pushRelation(rel);
+    }
+
+    // If the relation has already been seen in the current branch,
+    // it means there is a circular reference
+    if (relationsSeenInBranch[adjacentRelId]) {
+      throw new InvalidRelationError(
+        'A circular reference was found in the connect array. ' +
+          'One relation is trying to connect before/after another one that is trying to connect before/after it'
+      );
+    }
+
+    // Look if id is referenced elsewhere in the array
+    const adjacentRelIdx = firstSeen[adjacentRelId];
+
+    if (adjacentRelIdx) {
+      const adjacentRel = connectArr[adjacentRelIdx];
+
+      // Mark adjacent relation idx as computed,
+      // so it is not computed again later in the loop
+      computedIdx[adjacentRelIdx] = true;
+      computeRelation(adjacentRel, idx, { ...relationsSeenInBranch, [rel.id]: true });
+      pushRelation(rel);
+    } else if (strictSort) {
+      // If we reach this point, it means that the adjacent relation is not in the connect array
+      // and it is not in the database. This should not happen.
+      throw new InvalidRelationError(
+        `There was a problem connecting relation with id ${rel.id} at position ${JSON.stringify(
+          rel.position
+        )}. The relation with id ${adjacentRelId} needs to be connected first.`
+      );
+    } else {
+      // We are in non-strict mode so we can push the relation.
+      pushRelation({ id: rel.id, position: { end: true } });
+    }
+  };
+
+  // Iterate over connectArr and populate sortedConnect
+  connectArr.forEach((rel, idx) => {
+    computeRelation(rel, idx, {});
+  });
+
+  return sortedConnect;
```

Comment:
At first reading, I was completely lost, asking myself questions like "where is sortedConnect populated?", it appears that it is populated at line 48 by a function itself used at lines 60, 82 and 93, and returned at line 102. It doesn't feel instinctive to have to navigate the code like this to understand what it does. 😅

Reply:
Do you think this is more of code structure or not knowing the algorithm beforehand?(or both) Could it help with some more documentation?

Reply:
I personally find doing this in a more functional way makes code a little bit more bloated. But I am fully open to suggestions to improve this, I agree it's hard to read, but also the algo we found to fix the problem is hard too.